packages/next/lib/recursive-delete.ts

```diff
@@ -0,0 +1,60 @@
+import fs from 'fs'
+import { join } from 'path'
+import { promisify } from 'util'
+
+const readdir = promisify(fs.readdir)
+const stat = promisify(fs.stat)
+const rmdir = promisify(fs.rmdir)
+const unlink = promisify(fs.unlink)
+const sleep = promisify(setTimeout)
+
+const unlinkFile = async (p: string, t = 1): Promise<void> => {
+  try {
+    await unlink(p)
+  } catch (e) {
+    if ((e.code === 'EBUSY' || e.code === 'ENOTEMPTY' || e.code === 'EPERM' || e.code === 'EMFILE') && t < 3) {
+      await sleep(t * 100)
+      return unlinkFile(p, t++)
+    }
+
+    if (e.code === 'ENOENT') {
+      return
+    }
+
+    throw e
+  }
+}
+
+/**
+ * Recursively delete directory contents
+ * @param  {string} dir Directory to delete the contents of
+ * @param  {RegExp} filter Filter for the file name, only the name part is considered, not the full path
+ * @returns Promise void
+ */
+export async function recursiveDelete(dir: string, filter?: RegExp, ensure?: boolean): Promise<void> {
+  let result
+  try {
+    result = await readdir(dir)
+  } catch (e) {
+    if (e.code === 'ENOENT' && !ensure) return
+
+    throw e
+  }
+
+  await Promise.all(result.map(async (part: string) => {
```

Comment:
Do we want to use `Promise.all` here? We might eagerly bail while other promises are running.

We might want to wait for _all_ promises to reject or resolve using something like `p-reflect`: https://github.com/sindresorhus/p-reflect/blob/master/index.js